syntax = "proto2";

package dfproto.workdetailtest;

option optimize_for = LITE_RUNTIME;

message ProcessInfo {
    optional int32 pid = 1;
    optional uint64 cookie_address = 2;
    optional uint32 cookie_value = 3;
}

// GetProcessInfo: EmptyMessage -> ProcessInfo

enum Viewscreen {
    Other = 0;
    SetupDwarfGame = 1;
}

message GameState {
    optional uint64 world_loaded = 1;
    optional uint64 map_loaded = 2;
    optional Viewscreen viewscreen = 3;
}

// GetGameState: EmptyMessage -> GameState

message Result {
    optional bool success = 1;
    optional string error = 2;
}

message UnitId {
    optional uint32 id = 1;
}

enum UnitFlag {
    Unknown = 0;
    OnlyDoAssignedJobs = 1;
    AvailableForAdoption = 2;
    MarkedForSlaughter = 3;
    MarkedForGelding = 4;
}

message UnitFlagValue {
    optional UnitFlag flag = 1;
    optional bool value = 2;
}

message UnitProperties {
    // changes if set:
    optional string nickname = 1;
    repeated UnitFlagValue flags = 2;
}

message EditUnit {
    optional UnitId id = 1;
    optional UnitProperties changes = 2;
}

message EditUnits {
    repeated EditUnit units = 1;
}

message UnitFlagResult {
    optional UnitFlag flag = 1;
    optional Result result = 2;
}

message UnitResult {
    optional Result unit = 1;
    repeated UnitFlagResult flags = 2;
}

message UnitResults {
    repeated UnitResult results = 1;
}

// EditUnit: EditUnit -> UnitResult
// EditUnits: EditUnits -> UnitResults

enum WorkDetailMode {
    EverybodyDoesThis = 1;
    NobodyDoesThis = 2;
    OnlySelectedDoesThis = 3;
}

message WorkDetailAssignment {
    optional int32 unit_id = 1;
    optional bool enable = 2;
}

message WorkDetailLabor {
    optional int32 labor = 1;
    optional bool enable = 2;
}

message WorkDetailId {
    optional uint32 index = 1;
    optional string name = 2;
}

message WorkDetailProperties {
    // changes only if set or non-empty
    optional string name = 1;
    optional WorkDetailMode mode = 2;
    repeated WorkDetailAssignment assignments = 3;
    repeated WorkDetailLabor labors = 4;
    optional int32 icon = 5;
    optional bool no_modify = 6;
    optional bool cannot_be_everybody = 7;
}

message EditWorkDetail {
    optional WorkDetailId id = 1;
    optional WorkDetailProperties changes = 2;
}

message AddWorkDetail {
    optional uint32 position = 1; // push at end if missing
    optional WorkDetailProperties properties = 2;
}

message RemoveWorkDetail {
    optional WorkDetailId id = 1;
}

message MoveWorkDetail {
    optional WorkDetailId id = 1;
    optional uint32 new_position = 2;
}

message WorkDetailResult {
    optional Result work_detail = 1;
    optional Result mode = 2;
    repeated Result assignments = 3;
    repeated Result labors = 4;
    optional Result icon = 5;
}

// EditWorkDetail: EditWorkDetail -> WorkDetailResult
// AddWorkDetail: AddWorkDetail -> WorkDetailResult
// RemoveWorkDetail: RemoveWorkDetail -> Result
// MoveWorkDetail: MoveWorkDetail -> Result
